Sources/web3swift/Transaction/EIP2930Envelope.swift

```diff
@@ -0,0 +1,412 @@
+        toReturn += "s: " + String(self.s) + "\n"
+        return toReturn
+    }
+}
+
+extension EIP2930Envelope {
+    private enum CodingKeys: String, CodingKey {
+        case chainId
+        case nonce
+        case to
+        case value
+        case gasPrice
+        case gasLimit
+        case gas
+        case data
+        case input
+        case accessList
+        case v
+        case r
+        case s
+    }
+
+    public init?(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+
+        if !container.contains(.to) || !container.contains(.nonce) || !container.contains(.value) || !container.contains(.chainId) { return nil }
+        if !container.contains(.data) && !container.contains(.input) { return nil }
+        if !container.contains(.v) || !container.contains(.r) || !container.contains(.s) { return nil }
+
+        // everything we need is present, so we should only have to throw from here
+        self.internalChainID = try container.decodeHexIfPresent(to: BigUInt.self, key: .chainId) ?? 0
+        self.nonce = try container.decodeHex(to: BigUInt.self, key: .nonce)
+
+        let list = try? container.decode([AccessListEntry].self, forKey: .accessList)
+        self.accessList = list ?? []
+
+        let toString = try? container.decode(String.self, forKey: .to)
+        switch toString {
+        case nil, "0x", "0x0":
+            self.to = EthereumAddress.contractDeploymentAddress()
+        default:
+            // the forced unwrap here is safe as we trap nil in the previous case
+            // swiftlint:disable force_unwrapping
+            guard let ethAddr = EthereumAddress(toString!) else { throw Web3Error.dataError }
+            // swiftlint:enable force_unwrapping
+            self.to = ethAddr
+        }
+        self.value = try container.decodeHexIfPresent(to: BigUInt.self, key: .value) ?? 0
+        self.gasPrice = try container.decodeHexIfPresent(to: BigUInt.self, key: .gasPrice) ?? 0
+        self.gasLimit = try container.decodeHexIfPresent(to: BigUInt.self, key: .gas) ?? container.decodeHexIfPresent(to: BigUInt.self, key: .gasLimit) ?? 0
+
+        self.data = try container.decodeHexIfPresent(to: Data.self, key: .input) ?? container.decodeHex(to: Data.self, key: .data)
+        self.v = try container.decodeHex(to: BigUInt.self, key: .v)
+        self.r = try container.decodeHex(to: BigUInt.self, key: .r)
+        self.s = try container.decodeHex(to: BigUInt.self, key: .s)
+    }
+
+    // RLP encoding positions
+    fileprivate enum RlpKey: Int {
+        case chainId
+        case nonce
+        case gasPrice
+        case gasLimit
+        case destination
+        case amount
+        case data
+        case accessList
+        case sig_v
+        case sig_r
+        case sig_s
+        case total // not a real entry, used to auto-size based on number of keys
+    }
+
+    public init?(rawValue: Data) {
```

Comment:
One thing to keep in mind here is that while all types are currently RLP encoded, future types may not, which is why I avoided splitting this into two parts, a common generic part in `EIP2718Envelope`, and the type specific part here. And I would really like to avoid having any code in `EIP2718Envelope` that needs to know specifics of types derived from it... that is it should not contain any conditional execution based on `type`

I did consider having it unwrap the envelope and strip off the `type` byte, but saw no added benefit over doing it within each envelope, if anything it made things more confusing. However if there is a desire for that, I can implement it that way.

Reply:
I wanted to combine them as well, but in the end it is best if each final implementation is responsible for its own encoding and decoding, as that's where the differences lie, even if they are just subtle differences like the order or number of parameters.